### PR TITLE
undertow: Expose WebSocketChannel and WebSocketHttpExchange to the Camel Exchange

### PR DESCRIPTION
**Pull request by nhoughto**
This change adds the channel and undertow exchange to the camel exchange when available. This allows a whole set of new functionalities that are important for a proper websocket implementation.

For example this allows control over the underlying WebSocketChannel, if you want to close the channel for example, or check if it is still open. Passing the undertow exchange on socket `ON_OPEN` also gives the ability to access the initial transport headers, this could be used to authenticate a new socket for example.

As this change just adds the new values, in a cheap way, it didn't seem worth it to put it behind a configuration flag. As a consumer of this component it is likely what people expect anyway and aligns more with how other components behave.

### CODE REVIEW

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java [hunk 1]
```diff
@@ -198,14 +200,18 @@ private void sendResponse(HttpServerExchange httpExchange, Exchange camelExchang
...
      * @param message the message received via the {@link WebSocketChannel}
      */
-    public void sendMessage(final String connectionKey, final Object message) {
+    public void sendMessage(final String connectionKey, WebSocketChannel channel, final Object message) {
 
         final Exchange exchange = getEndpoint().createExchange();
 
         // set header and body
         exchange.getIn().setHeader(UndertowConstants.CONNECTION_KEY, connectionKey);
+        if(channel != null) {
```
Comment (oscerd) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: Please run with the sourcecheck profile enabled and fix the code style.

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java [hunk 2]
```diff
@@ -223,16 +229,23 @@ public void done(boolean doneSync) {
...
-    public void sendEventNotification(String connectionKey, EventType eventType) {
+    public void sendEventNotification(String connectionKey, WebSocketHttpExchange transportExchange, WebSocketChannel channel, EventType eventType) {
         final Exchange exchange = getEndpoint().createExchange();
 
         final Message in = exchange.getIn();
         in.setHeader(UndertowConstants.CONNECTION_KEY, connectionKey);
         in.setHeader(UndertowConstants.EVENT_TYPE, eventType.getCode());
         in.setHeader(UndertowConstants.EVENT_TYPE_ENUM, eventType);
-
+        if(channel != null){
```
Comment (oscerd) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: ditto

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java [hunk 3]
```diff
@@ -223,16 +229,23 @@ public void done(boolean doneSync) {
...
+            in.setHeader(UndertowConstants.CHANNEL, channel);
+        }
+        if(transportExchange != null){
```
Comment (oscerd) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: ditto

Comment (nhoughto) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: done

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java (hunk 2 above)
Comment (nhoughto) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: done

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java (hunk 1 above)
Comment (nhoughto) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java]: done

**Comment by nhoughto** [comment]
codestyle fixed, is there anything else needed to backport to pre 3.x versions?

**Comment by oscerd** [comment]
To backport on camel-2.x branch you'll need to check carefully, because on master we did an heavy refactoring. Since this is something new, it will be only available in 2.25.0

**Comment by nhoughto** [comment]
So do I need to do something extra for it to land in 2.25? or is this enough?

**Comment by oscerd** [comment]
You can create a Pull Request against camel-2.x

**Comment by nhoughto** [comment]
excellent, will do thanks

**Comment by oscerd** [comment]
I'll merge this one tomorrow

**Comment by oscerd** [comment]
Thanks

**Pull request closed without merging**